Add option "buffer_flush_interval"

abicky opened the pull request:
If metrics are recorded infrequently, it might takes much time until the message buffer flushes them and we can see them on Datadog. This option is useful in such a case.

djmitche commented:
I think this makes a lot of sense, and I appreciate that it's optional, so the overhead of an additional thread is only borne by those who want to use it.  I know this is old.. do you want to rebase onto the latest changes?

mopp commented:
We faced the same issue in our application.
This change looks good 👍 

@abicky do you want to rebase this PR?

abicky commented:
@djmitche @mopp Thank you for your comments and I'm sorry for my late reply. I've rebased the branch. At first, I implemented the feature in `Datadog::Statsd::Sender`, but now that another sender, `Datadog::Statsd::SingleThreadSender`, exists, I've reimplemented it in `Datadog::Statsd::MessageBuffer` to share the logic between the senders.

djmitche commented on lib/datadog/statsd/message_buffer.rb:
```diff
@@ -17,9 +18,15 @@ def initialize(connection,
         @max_pool_size = max_pool_size
         @overflowing_stategy = overflowing_stategy
+        @flush_interval = flush_interval
+        @mon = Monitor.new
```
It would be good to have a comment as to what this monitor protects

abicky replied:
I've written a comment in the commit https://github.com/DataDog/dogstatsd-ruby/pull/154/commits/a48591503ed5e57ba3fc882c1bba7b89c375b975.

djmitche commented on lib/datadog/statsd/message_buffer.rb:
```diff
@@ -92,6 +123,33 @@ def ensure_sendable!(message_size)
+          @mon.synchronize do
+            @cv.wait(flush_interval - (current_time - last_flush_time))
+            last_flush_time = current_time
+            flush
```
`flush` also uses `@mon.synchronize`, so I'd expect this to deadlock (unless Monitor is re-entrant, but I don't see that in its API docs).

abicky replied:
Yes, this is why I use `Monitor` instead of `Mutex`. `Monitor` is re-entrant.
I couldn't find the official document written in English, but [the document written in Japanese](https://docs.ruby-lang.org/ja/3.1/library/monitor.html) states that it provides the feature allowing the same thread to lock multiple times:

> monitor は以下のような Thread::Mutex としての機能も提供します。
> 
> * lock の持ち主がスレッドである Mutex / 何度も lock できる Mutex
>    - lock したスレッドを Mutex 側が覚えていて
>    - そのスレッドがもう一度 lock しようとしてもブロックしない
>    - synchronize は lock していなかったら通常どおり、 自分が lock していたら ただ yield するだけ (lock/unlockもしない)
>    - unlock はそのスレッドだけができる

abicky replied:
You can try the behavior with the following script:

```ruby
require 'monitor'

mon = Monitor.new
mon.synchronize do
  mon.synchronize do
    puts 'hello'
  end
end
```

djmitche replied:
OK, that's a bit bizarre but I'm glad it works :)

djmitche commented on lib/datadog/statsd/message_buffer.rb:
```diff
@@ -92,6 +123,33 @@ def ensure_sendable!(message_size)
+
+      def create_flush_thread
+        flush_thread = Thread.new(&method(:flush_loop))
+        flush_thread.name = 'Statsd MessageBuffer' if flush_thread.respond_to?(:name=)
```
For better or for worse, the way we've done this in other files is e.g.,
```ruby
@sender_thread.name = "Statsd Sender" unless Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.3')
```
so let's stick to that pattern.

abicky replied:
I've changed the code to follow the same way in the commit https://github.com/DataDog/dogstatsd-ruby/pull/154/commits/f5f8686b1161aced41df61e95a4744559cfccab2.

djmitche commented on lib/datadog/statsd/message_buffer.rb:
```diff
@@ -92,6 +123,33 @@ def ensure_sendable!(message_size)
+        end
+      end
+
+      if Process.const_defined?(:CLOCK_MONOTONIC)
+        def current_time
+          Process.clock_gettime(Process::CLOCK_MONOTONIC)
+        end
+      else
+        def current_time
+          Time.now
+        end
+      end
```
`Statsd.now` will do this for you :)

abicky replied:
I didn't know that! It seems a private instance method, so I have to make it a public singleton method to use it from here. I'm afraid that some users call it from their code, that is, outside this library if I make it public. What do you think about that?

djmitche replied:
Oh, hm, that's a good point.  Let's leave it as-is, then.

djmitche commented on lib/datadog/statsd/message_buffer.rb:
```diff
@@ -28,30 +35,53 @@ def add(message)
+        @mon.synchronize {
+          clear_buffer
+          connection.reset_telemetry
+          @flush_thread = create_flush_thread if @flush_interval
```
This changes the semantics of `reset` such that it can only be called when the buffer is in a closed state (has no running flush thread).  As currently used in `Sender`, that's not quite true: it's called when the buffer is not closed, but coincidentally in a condition where it's likely that the flush thread has been killed by a `fork`.

Maybe let's do a similar thing, where the flush thread is only created if it's `nil` or if it's not alive, similar to
https://github.com/DataDog/dogstatsd-ruby/blob/48f99e869e969431db3264dd50f1bcd7125de919/lib/datadog/statsd/sender.rb#L64-L64

abicky replied:
I think you suggest I write as follows, but I thought the change like https://github.com/DataDog/dogstatsd-ruby/pull/154/commits/89811077faf05e0f16676a0e4f94dcff338019ac would be better:

```ruby
if @flush_interval && (@flush_thread.nil? || @flush_thread.stop?)
  @flush_thread = create_flush_thread
end
```

You can refer to the commit message for more details. How about this?

djmitche replied:
That makes great sense - thank you!

djmitche reviewed (changes requested):
Looking good!  I'm fairly new to Ruby so if I've misunderstood Monitor, please let me know.

abicky commented:
@djmitche Thank you for your review! I made some changes. Can you take a look?

djmitche reviewed (approved):
Looks great to me!  I'd like @remeh to have a look as well before we merge.

remeh commented on lib/datadog/statsd/message_buffer.rb:
```diff
@@ -28,30 +39,53 @@ def add(message)
         return nil unless ensure_sendable!(message_size)
 
-        flush if should_flush?(message_size)
+        @mon.synchronize {
```
`add` is in the hot-path of the library when running at high-throughput, and, while I agree we need some synchronization here (i.e. [I have a draft with a simpler multi-threaded Sender code, and it needs a `Mutex` in `add` as well](https://github.com/DataDog/dogstatsd-ruby/pull/209/files#diff-22ef838ef4e2984fdedbeae928e214979fa440026ddc2f3f96b90328098b5ee8R61)), I'm wondering if using a `Monitor` is the best thing to do here, as:
  * it can be slightly more confusing than a mutex since Monitor are reentrant ([see this post](https://japgolly.blogspot.com/2012/04/ruby-mutex-reentrancy.html))
  * could be slower than a mutex

abicky replied:
I think it is better to use `Monitor` for the following reasons:

* It has better performance than `Mutex`
    - As described on [the release note of Ruby 2.7.0](https://www.ruby-lang.org/en/news/2019/12/25/ruby-2-7-0-released/), the performance of `Monitor` has been improved. (cf. https://github.com/ruby/ruby/commit/caac5f777ae288b5982708b8690e712e1cae0cf6)
* It makes code clear
    - We don't have to manage who is the owner of the lock
* It is also used in [Fluentd](https://github.com/fluent/fluentd), which also processes so many events

Actually, the following benchmark shows `Monitor` has better performance than `Mutex`. Of course, the lock makes the performance worse than before, but I think `MessageBuffer` still has good performance.

```ruby
require 'monitor'
require 'benchmark/ips'

class MessageBufferBase
  PAYLOAD_SIZE_TOLERANCE = 0.05

  def initialize
    @max_payload_size = 1_432
    @max_pool_size = Float::INFINITY
    @overflowing_stategy = :drop

    @buffer = String.new
    clear_buffer

    @closed = false
  end

  private

  attr :max_payload_size
  attr :max_pool_size

  attr :overflowing_stategy
  attr_reader :flush_interval

  attr :buffer

  def _flush
    return if buffer.empty?

    clear_buffer
  end

  def should_flush?(message_size)
    return true if buffer.bytesize + 1 + message_size >= max_payload_size

    false
  end

  def clear_buffer
    buffer.clear
    @message_count = 0
  end

  def preemptive_flush?
    @message_count == max_pool_size || buffer.bytesize > bytesize_threshold
  end

  def ensure_sendable!(message_size)
    return true if message_size <= max_payload_size

    if overflowing_stategy == :raise
      raise Error, 'Message too big for payload limit'
    end

    false
  end

  def bytesize_threshold
    @bytesize_threshold ||= (max_payload_size - PAYLOAD_SIZE_TOLERANCE * max_payload_size).to_i
  end
end

class MessageBufferWithoutLock < MessageBufferBase
  def add(message)
    message_size = message.bytesize

    return nil unless message_size > 0 # to avoid adding empty messages to the buffer
    return nil unless ensure_sendable!(message_size)

    raise Error, 'buffer is closed' if @closed
    flush if should_flush?(message_size)

    buffer << "\n" unless buffer.empty?
    buffer << message

    @message_count += 1

    # flush when we're pretty sure that we won't be able
    # to add another message to the buffer
    flush if preemptive_flush?

    true
  end

  def flush
    _flush
  end
end

class MessageBufferWithMutex < MessageBufferBase
  def initialize
    super
    @mutex = Mutex.new
    @mutex_owner = nil
  end

  def add(message)
    message_size = message.bytesize

    return nil unless message_size > 0 # to avoid adding empty messages to the buffer
    return nil unless ensure_sendable!(message_size)

    @mutex.synchronize {
      begin
        @mutex_owner = Thread.current
        raise Error, 'buffer is closed' if @closed
        flush if should_flush?(message_size)

        buffer << "\n" unless buffer.empty?
        buffer << message

        @message_count += 1

        # flush when we're pretty sure that we won't be able
        # to add another message to the buffer
        flush if preemptive_flush?
      ensure
        @mutex_owner = nil
      end
    }

    true
  end

  def flush
    if @mutex_owner != Thread.current
      @mutex.lock
      locked = true
    end
    _flush
  ensure
    @mutex.unlock if locked
  end
end

class MessageBufferWithMonitor < MessageBufferBase
  def initialize
    super
    @mon = Monitor.new
  end

  def add(message)
    message_size = message.bytesize

    return nil unless message_size > 0 # to avoid adding empty messages to the buffer
    return nil unless ensure_sendable!(message_size)

    @mon.synchronize {
      raise Error, 'buffer is closed' if @closed
      flush if should_flush?(message_size)

      buffer << "\n" unless buffer.empty?
      buffer << message

      @message_count += 1

      # flush when we're pretty sure that we won't be able
      # to add another message to the buffer
      flush if preemptive_flush?
    }

    true
  end

  def flush
    @mon.synchronize { _flush }
  end
end


Benchmark.ips do |x|
  b1 = MessageBufferWithoutLock.new
  b2 = MessageBufferWithMutex.new
  b3 = MessageBufferWithMonitor.new

  messages = [
    "a" * 20,
    "a" * 30,
    "a" * 40,
  ]

  x.report('MessageBufferWithoutLock#add') { messages.each { |m| b1.add(m) } }
  x.report('MessageBufferWithMutex#add') { messages.each { |m| b2.add(m) } }
  x.report('MessageBufferWithMonitor#add') { messages.each { |m| b3.add(m) } }
  x.compare!
end
```

The result in my environment (macOS 11.6.1, 2.3 GHz Quad-Core Intel Core i7, ruby 3.0.2p107, ) is here:

```
Warming up --------------------------------------
MessageBufferWithoutLock#add
                        57.506k i/100ms
MessageBufferWithMutex#add
                        42.935k i/100ms
MessageBufferWithMonitor#add
                        38.971k i/100ms
Calculating -------------------------------------
MessageBufferWithoutLock#add
                        566.378k (± 2.4%) i/s -      2.875M in   5.079683s
MessageBufferWithMutex#add
                        417.459k (± 7.1%) i/s -      2.104M in   5.080333s
MessageBufferWithMonitor#add
                        415.527k (± 3.5%) i/s -      2.104M in   5.070831s

Comparison:
MessageBufferWithoutLock#add:   566377.6 i/s
MessageBufferWithMutex#add:   417458.5 i/s - 1.36x  (± 0.00) slower
MessageBufferWithMonitor#add:   415527.0 i/s - 1.36x  (± 0.00) slower

```

abicky replied:
In the benchmark, I don't consider the flush thread, because I assume that it calls `flush` much less frequently than `add`, which is called by the sender thread.

abicky replied:
If you prefer Mutex, I'll change the code in the same way as the benchmark.

remeh replied:
Thanks for the benchs @abicky and having took the time to share your thoughts, that's very convincing 💯 

I'll do a second complete review ASAP @djmitche with process forking in mind while reviewing (forks are heavily occurring for people using Puma, Sidekiq, etc...).

remeh replied:
Are we sure that these locks are per process and not system-wide with a monitor? I guess so but just to make sure.
Since workers libraries rely on forks, this lock can't be system-wide for our usage.

djmitche replied:
If I understand @remeh, the issue you're concerned with is that the main thread calls `fork()` while another thread is in the midst of executing an `add` call and is holding the lock.  In the child process, that thread would not exist but the lock would still be held, so subsequent calls to `add` would hang.

abicky replied:
I'm sorry for my late reply. 

> Are we sure that these locks are per process and not system-wide with a monitor?

Yes. As you can see below, the monitor in the child process is unlocked:

```ruby
require 'monitor'

mon = Monitor.new
thread = Thread.new do
  loop do
    puts "pid: #{Process.pid}, mon.mon_locked?: #{mon.mon_locked?}"
    sleep(1)
  end
end
mon.mon_enter

child_pid = fork do
  puts "child process. pid: #{Process.pid}, thread.alive?: #{thread.alive?}, mon.mon_locked?: #{mon.mon_locked?}"
  mon.mon_enter
  puts "sleep a little"
  sleep(3)
end

puts "parent process. pid: #{Process.pid}, child pid: #{child_pid}, thread.alive?: #{thread.alive?}, mon.mon_locked?: #{mon.mon_locked?}"

Process.waitpid(child_pid)
```

The output is here:

```
parent process. pid: 46387, child pid: 46417, thread.alive?: true, mon.mon_locked?: true
pid: 46387, mon.mon_locked?: true
child process. pid: 46417, thread.alive?: false, mon.mon_locked?: true
sleep a little
pid: 46387, mon.mon_locked?: true
pid: 46387, mon.mon_locked?: true
```

abicky replied:
By the way, I've implemented the feature in another way https://github.com/DataDog/dogstatsd-ruby/pull/231. I'd like you to merge one you prefer and close the other.

djmitche commented:
It seems like we all prefer #231, so I'll close this PR.
